### PR TITLE
haechi/13

### PR DESCRIPTION
**Pull request by tommy4241**
#96

### CODE REVIEW

File: test/Haechi/13.sol [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        uint256 amt = 1000000;
+        // mint token to trader
+        deal(address(tokens[1]), address(accounts[1]), amt * decimals[1]);
+
+        uint256 eurocToSend = 1000000;
+        // destroy balance by first sending 100:1 Usdc : Euroc ratio
+        deal(address(tokens[1]), address(curves[1]), eurocToSend);
+        deal(address(tokens[3]), address(curves[1]), eurocToSend * 100);
+
+        uint256 forexBalance = tokens[1].balanceOf(address(accounts[1]));
```
Comment (kafitz) [inline test/Haechi/13.sol:174]:
```suggestion
```

File: src/assimilators/AssimilatorV2.sol [hunk 2]
```diff
@@ -89,23 +89,23 @@ contract AssimilatorV2 is IAssimilator {
...
-        uint256 _usdcBal = usdc.balanceOf(_addr).mul(1e18).div(_quoteWeight);
-
-        // Rate is in 1e6
-        uint256 _rate = _usdcBal.mul(10**tokenDecimals).div(_tokenBal);
-
-        amount_ = (_amount.mulu(10**tokenDecimals) * 1e6) / _rate;
-
+        amount_ =
+            (_amount.mulu((1e6)) * _quoteWeight * token.balanceOf(_addr)) /
+            (usdc.balanceOf(_addr) * _baseWeight);
```
Comment (chanharryhk) [inline src/assimilators/AssimilatorV2.sol:103]: Why did you change the `amount_` value here? Shouldn't it be left as the same as before?

File: test/Haechi/13.sol [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        uint256 amt = 1000000;
+        // mint token to trader
+        deal(address(tokens[1]), address(accounts[1]), amt * decimals[1]);
+
+        uint256 eurocToSend = 1000000;
+        // mint tokens to attacker
+        deal(address(tokens[1]), address(curves[1]), eurocToSend);
+        deal(address(tokens[3]), address(curves[1]), eurocToSend * 100);
+
+        uint256 forexBalance = tokens[1].balanceOf(address(accounts[1]));
```
Comment (kafitz) [inline test/Haechi/13.sol:138]:
```suggestion
```

File: test/Haechi/13.sol [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+    function testMinDeposit() public {
+        uint256 amt = 1000000;
+        // mint token to trader
+        deal(address(tokens[1]), address(accounts[1]), amt * decimals[1]);
+
+        uint256 eurocToSend = 1000000;
+        // mint tokens to attacker
+        deal(address(tokens[1]), address(curves[1]), eurocToSend);
+        deal(address(tokens[3]), address(curves[1]), eurocToSend * 100);
```
Comment (kafitz) [inline test/Haechi/13.sol:136]: Here we're actually seeding the LP, right?

File: test/Haechi/13.sol [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        uint256 forexBalance = tokens[1].balanceOf(address(accounts[1]));
+
+        cheats.startPrank(address(accounts[1]));
+        tokens[1].approve(address(curves[1]), type(uint256).max);
+        tokens[3].approve(address(curves[1]), type(uint256).max);
+        cheats.stopPrank();
+        // destroy balance by first sending 100:1 Usdc : Euroc ratio
+
+        uint256 minEurocToDeposit = 500000000000000;
+        uint256 minUsdcToDeposit = 350000000000000;
```
Comment (kafitz) [inline test/Haechi/13.sol:147]:
```suggestion
```
Unneeded here

**Review by kafitz — APPROVED** [review]
Overall I'd just suggest to tidy up the test, but LGTM

**Review by chanharryhk — APPROVED** [review]
Just 1 question but looks good to me otherwise.

File: test/Haechi/13.sol (hunk 4 above)
Comment (tommy4241) [reply test/Haechi/13.sol:136]: it's a bit different from seeding what we used to do after new pool deployment. This time, we are directly sending tokens to the pool with 100:1 ratio and this doesn't mint any LPTs.

File: src/assimilators/AssimilatorV2.sol (hunk 2 above)
Comment (tommy4241) [reply src/assimilators/AssimilatorV2.sol:103]: this isn't related to front-running. I've tried to simply the function body.

File: test/Haechi/13.sol (hunk 4 above)
Comment (kafitz) [reply test/Haechi/13.sol:136]: Ah ok I see, the comment threw me off. Bit of a nitpick, but we should can change or delete it if it's not accurate. Just a matter of safety for naive folks like me looking at the code and perhaps assuming the wrong thing.

File: src/assimilators/AssimilatorV2.sol (hunk 2 above)
Comment (chanharryhk) [reply src/assimilators/AssimilatorV2.sol:103]: Nitpick but I would change it back to improve the clarity of the code. The haechi guys dont have to reread to ensure that this part is correct and mainly focus on the minquote and minbase
